File: platform/src/main/java/ru/abondin/hreasy/platform/auth/AuthHandler.java
package ru.abondin.hreasy.platform.auth;

import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.security.authentication.ReactiveAuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.context.ReactiveSecurityContextHolder;
import org.springframework.security.core.context.SecurityContext;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.web.context.HttpSessionSecurityContextRepository;
import org.springframework.stereotype.Component;
import org.springframework.web.server.WebSession;
import reactor.core.publisher.Mono;
import ru.abondin.hreasy.platform.sec.UserDetailsWithEmployeeInfo;

import java.util.ArrayList;
import java.util.Collection;
import java.util.stream.Collectors;

/**
 * Helper for auth
 */
@Component
@RequiredArgsConstructor
@Slf4j
public class AuthHandler {

    @Data
    @NoArgsConstructor
    public static class CurrentUserDto {
        private String username;
        private String email;
        private Collection<String> authorities = new ArrayList<>();
        private AuthContext.EmployeeInfo employee;

        public CurrentUserDto(AuthContext authContext) {
            this.username = authContext.getUsername();
            this.email = authContext.getEmail();
            this.authorities = authContext.getAuthorities();
            if (authContext.getEmployeeInfo() != null) {
                this.employee = new AuthContext.EmployeeInfo(authContext.getEmployeeInfo());
            }
        }
    }

    @Data
    @AllArgsConstructor
    @NoArgsConstructor
    public static class LoginResponse {
        private CurrentUserDto currentUser;
    }

    @Data
    @AllArgsConstructor
    @NoArgsConstructor
    public static class LogoutResponse {
        private String status = "OK";
    }

    public final static Mono<AuthContext> currentAuth() {
        return ReactiveSecurityContextHolder.getContext()
                .map(it -> buildFromSecurityContext(it));
    }

    private static AuthContext buildFromSecurityContext(SecurityContext context) {
        var userDetails = (UserDetails) context.getAuthentication().getPrincipal();
        var email = userDetails.getUsername();
        AuthContext.EmployeeInfo employee;
        if (userDetails instanceof UserDetailsWithEmployeeInfo && ((UserDetailsWithEmployeeInfo) userDetails).getEmployeeId() != null) {
            var withEmployeeInfo = ((UserDetailsWithEmployeeInfo) userDetails);
            employee = new AuthContext.EmployeeInfo(withEmployeeInfo.getEmployeeId(),
                    withEmployeeInfo.getDepartmentId(),
                    withEmployeeInfo.getCurrentProjectId(),
                    withEmployeeInfo.getAccessibleDepartments(),
                    withEmployeeInfo.getAccessibleBas(),
                    withEmployeeInfo.getAccessibleProjects(),
                    withEmployeeInfo.getLoggedInType());
        } else {
            employee = null;
        }
        var ctx = new AuthContext(userDetails.getUsername(), email,
                userDetails.getAuthorities().stream().map(GrantedAuthority::getAuthority).collect(Collectors.toList()),
                employee);
        log.debug("Auth context built for {}", userDetails.getUsername());
        return ctx;
    }

    private final ReactiveAuthenticationManager authenticationManager;

    public Mono<LoginResponse> login(UsernamePasswordAuthenticationToken authenticationToken, WebSession webSession) {
        return authenticationManager.authenticate(authenticationToken).flatMap(
                authResult -> ReactiveSecurityContextHolder.getContext().map(
                        securityContext -> {
                            securityContext.setAuthentication(authResult);
                            return securityContext;
                        }).contextWrite(ReactiveSecurityContextHolder.withAuthentication(authResult))
        ).flatMap(securityContext -> {
            webSession.getAttributes().put(HttpSessionSecurityContextRepository.SPRING_SECURITY_CONTEXT_KEY, securityContext);
            return webSession.save();
        }).then(currentAuth()).map(
                auth -> new LoginResponse(new CurrentUserDto(auth))
        );
    }

    public Mono<AuthContext> login(UsernamePasswordAuthenticationToken authenticationToken) {
        return authenticationManager.authenticate(authenticationToken).flatMap(
                authResult ->
                        ReactiveSecurityContextHolder.getContext().map(
                                securityContext -> {
                                    securityContext.setAuthentication(authResult);
                                    return securityContext;
                                }
                        ).contextWrite(ReactiveSecurityContextHolder.withAuthentication(authResult)).map(
                                context -> buildFromSecurityContext(context)));
    }

    public Mono<LogoutResponse> logout(WebSession webSession) {
        webSession.getAttributes().remove(HttpSessionSecurityContextRepository.SPRING_SECURITY_CONTEXT_KEY);
        return webSession.invalidate().map(result -> new LogoutResponse());
    }
}

File: platform/src/main/kotlin/ru/abondin/hreasy/platform/sec/AuthHandler.kt
package ru.abondin.hreasy.platform.sec

import org.springframework.security.authentication.ReactiveAuthenticationManager
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.context.ReactiveSecurityContextHolder
import org.springframework.security.core.context.SecurityContext
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.web.context.HttpSessionSecurityContextRepository
import org.springframework.stereotype.Component
import org.springframework.web.server.WebSession
import reactor.core.publisher.Mono
import ru.abondin.hreasy.platform.config.AuthContext
import ru.abondin.hreasy.platform.config.HrEasySecurityProps

/**
 * Helper for auth
 */
@Component
class AuthHandler(private val authenticationManager: ReactiveAuthenticationManager,
                  private val securityProps: HrEasySecurityProps) {


    fun login(authenticationToken: UsernamePasswordAuthenticationToken, webSession: WebSession): Mono<LoginResponse> {
        return authenticationManager.authenticate(authenticationToken).flatMap { authResult ->
            ReactiveSecurityContextHolder.getContext().map { securityContext ->
                securityContext.authentication = authResult;
                return@map securityContext;
            }.subscriberContext(ReactiveSecurityContextHolder.withAuthentication(authResult))
        }.flatMap { securityContext ->
            webSession.attributes.put(HttpSessionSecurityContextRepository.SPRING_SECURITY_CONTEXT_KEY, securityContext)
            return@flatMap webSession.save()
        }.then(
                currentAuth()
        ).map { auth ->
            LoginResponse(CurrentUserDto(auth))
        };
    }

    fun login(authenticationToken: UsernamePasswordAuthenticationToken): Mono<AuthContext> {
        return authenticationManager.authenticate(authenticationToken).flatMap { authResult ->
            ReactiveSecurityContextHolder.getContext().map { securityContext ->
                securityContext.authentication = authResult;
                return@map securityContext;
            }.subscriberContext(ReactiveSecurityContextHolder.withAuthentication(authResult))
        }.map { context ->
            buildFromSecurityContext(context);
        }
    }

    fun logout(webSession: WebSession): Mono<LogoutResponse> {
        webSession.getAttributes().remove(HttpSessionSecurityContextRepository.SPRING_SECURITY_CONTEXT_KEY);
        return webSession.invalidate().map { LogoutResponse() };
    }

    /**
     * Get current auth info from Spring Reactive Security Context
     */
    companion object {
        //FIXME Get from securityProps.emailSuffix
        val emailSuffix = "@stm-labs.ru";

        fun currentAuth(): Mono<AuthContext> =
                ReactiveSecurityContextHolder
                        .getContext()
                        .map { buildFromSecurityContext(it) }

        /**
         * Get employee email from username
         */
        fun emailFromUsername(username: String): String {
            return "${username}${emailSuffix}";
        }

        private fun buildFromSecurityContext(context: SecurityContext): AuthContext {
            val userDetails = context.authentication.principal as UserDetails;
            val email = emailFromUsername(userDetails.username);
            val employee: AuthContext.EmployeeInfo?;
            employee = if (userDetails is UserDetailsWithEmployeeInfo && userDetails.employeeId != null) {
                AuthContext.EmployeeInfo(userDetails.employeeId);
            } else {
                null;
            }
            return AuthContext(userDetails.username, email, userDetails.authorities.map { a -> a.authority }, employee);
        }
    }


}


data class LoginResponse(
        val currentUser: CurrentUserDto
);
data class LogoutResponse(
        val status: String = "OK"
);

data class CurrentUserDto(
        val username: String,
        val email: String,
        val authorities: Collection<String>,
        val employee: EmployeeInfo?
) {
    constructor(authContext: AuthContext) : this(
            authContext.username,
            authContext.email,
            authContext.authorities,
            if (authContext.employeeInfo == null) null else EmployeeInfo(authContext.employeeInfo)
    )

    data class EmployeeInfo(val id: Int) {
        constructor(employeeInfo: AuthContext.EmployeeInfo) : this(employeeInfo.id);
    }
}